server/src/service.rs

```diff
@@ -458,4 +474,42 @@ impl Service {
 
         Ok(signal_emitter)
     }
+
+    pub async fn remove_client(&self, old_owner: &str) -> Option<OwnedObjectPath> {
+        self.clients.lock().await.remove(old_owner)
+    }
+
+    pub async fn cleanup_sessions_on_client_disconnect(&self) -> Result<(), Error> {
+        let rule = zbus::MatchRule::builder()
+            .msg_type(zbus::message::Type::Signal)
+            .sender("org.freedesktop.DBus")?
+            .member("NameOwnerChanged")?
+            .build();
+        let mut stream = zbus::MessageStream::for_match_rule(rule, &self.connection, None).await?;
+
+        while let Some(message) = stream.try_next().await? {
+            let Ok((_name, old_owner, new_owner)) =
+                message.body().deserialize::<(String, String, String)>()
+            else {
+                continue;
+            };
+
+            if new_owner.is_empty() {
+                if let Some(session_path) = self.remove_client(&old_owner).await {
+                    if let Some(session) = self.session(&session_path).await {
+                        match session.close(self.object_server()).await {
```

Comment:
this is weird, instead remove the object_server param and make that close function use the object server from self.service.object_server or so.

Reply:
Done, thanks!